### PR TITLE
Do not call `check_expr` in `check_compatible`, since it has side-effects

### LINKED ISSUE #98894: Calling a function with wrong number of arguments and a `while`  loop or its equivalence yields weird diagnostics
## Explanation / examples

This perfectly wrong statement causes compiler to mention `if`s, `!` and `()` even though nothing like this is written in the code:
```rust
(|_, ()| ())(while true {});
```
```text
error[E0308]: `if` and `else` have incompatible types
 --> ./t.rs:2:18
  |
2 |     (|_, ()| ())(while true {});
  |                  ^^^^^^^^^^^--
  |                  |          |
  |                  |          expected because of this
  |                  expected `()`, found `!`
  |
  = note: expected unit type `()`
                  found type `!`
```
(here and after normal errors about argument number/type mismatch are not shown as they are not the problem)

Originally I've found this while editing some async code and passing `String` instead of `&str`. I was very confused and thought that it somehow relates to `async`/`impl Trait`/etc but after some minification I was left with the snippet above.

This error becomes less confusing after desugaring `while` loop by hand:
```rust
(|_, ()| ())(loop { if true {} else {break;} });
```
```
error[E0308]: `if` and `else` have incompatible types
 --> ./t.rs:2:42
  |
2 |     (|_, ()| ())(loop { if true {} else {break;} });
  |                                 --       ^^^^^^ expected `()`, found `!`
  |                                 |
  |                                 expected because of this
  |
  = note: expected unit type `()`
                  found type `!`
```
Note that using `break;` is _required_ to trigger this bug, using `break` doesn't trigger it for some reason.

In fact, you don't even need a `loop`:
```rust
(|_, ()| ())(if true {} else {return;});
```
```
error[E0308]: `if` and `else` have incompatible types
 --> ./t.rs:2:35
  |
2 |     (|_, ()| ())(if true {} else {return;});
  |                          --       ^^^^^^^ expected `()`, found `!`
  |                          |
  |                          expected because of this
  |
  = note: expected unit type `()`
                  found type `!`
```

So in the end, this can be triggered with an `if`/`else` one of branches on which is diverging expression with `;` that is being passed to a function with wrong number of arguments.

## Note about confusion / applicability to real programs

As I previously said, I've found this while working on async code and I think this is the main way how this error can be seen (let's be serious, nobody calls functions with `while` loop arguments). The reason why it's easier to find this error with async is that `async{}` blocks are commonly passed to functions (like `tokio::spawn`) and can contain `while` loops inside them.

In my case the code looked something like this:

```rust
spawn(arg_with_wrong_type, async {
    while let Some(_) = rx.recv().await {
        // ..

        some_atomic.store(true, Ordering::Relaxed);
    }
})
```

Which caused an error like this:

Note that "expected because of this" points to the last statement in the `while{}` block even though it has nothing to do, really, with the error.

## Bisection

searched nightlies: from nightly-2022-04-10 to nightly-2022-05-09
regressed nightly: nightly-2022-04-17
searched commit range: https://github.com/rust-lang/rust/compare/3f391b84552f210adec7893b50c5da74f9362ae4...878c7833f6c1ff10e2fd89074e5bd4ef5ff15936
regressed commit: https://github.com/rust-lang/rust/commit/07bb916d44a66d2caba427c7ee132bbeb245977b

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.3</summary>

Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --script=./test.sh --end=2022-05-09
```

`src/main.rs`:
```rust
fn main() {
    (|_, ()| ())(async { while true {} });
}
```
`test.sh`:
```sh
#!/bin/sh

# Fail if we no longer get a single error (and instead get at least two)
cargo check 2>&1 | grep "due to previous error;"
```

</details>

## Cause

Bisection shows that this issue was introduced by #92364 which aligns with other observations. This is likely caused by the same underlying problem as #96335: diagnostic code calls `check_expr` a second time which confuses different parts of the compiler.

cc @compiler-errors as they've worked on fixing issues caused by #92364 before

## Meta

This is reproducible with `1.63.0-beta.3` (2022-07-02 59f577d6388060f65b5f) and with most nightly-es after `nightly-2022-04-17` (except for those that ICE).

@rustbot label +A-diagnostics +D-confusing +D-incorrect

### LINKED ISSUE #98897: Passing array of `return` to a function with wrong number/type of arguments causes an ICE
Found this one while researching for #98894, probably has a related cause (since #92364 diagnostics call `check_expr` a second time which confuses the compiler). This is an ICE though so opening a dedicated issue.

cc @compiler-errors 

### Code

```Rust
(|_, ()| ())([return, ()]);
```

### Bisection

searched nightlies: from nightly-2022-04-10 to nightly-2022-05-09
regressed nightly: nightly-2022-04-17
searched commit range: https://github.com/rust-lang/rust/compare/3f391b84552f210adec7893b50c5da74f9362ae4...878c7833f6c1ff10e2fd89074e5bd4ef5ff15936
regressed commit: https://github.com/rust-lang/rust/commit/07bb916d44a66d2caba427c7ee132bbeb245977b

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.3</summary>

Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --regress=ice --end=2022-05-09
```

`src/main.rs`:
```rust
fn main() {
    (|_, ()| ())([return, ()]);
}
```

</details>

### Meta

`rustc --version --verbose`:
```
rustc 1.64.0-nightly (20a6f3a8a 2022-06-25)
binary: rustc
commit-hash: 20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753
commit-date: 2022-06-25
host: x86_64-unknown-linux-gnu
release: 1.64.0-nightly
LLVM version: 14.0.5
```

### Error output

```
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `WarnedAlways`,
 right: `Maybe`', compiler/rustc_typeck/src/check/expr.rs:1309:13
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
thread 'rustc' panicked at 'assertion failed: `(left == right)`
  left: `WarnedAlways`,
 right: `Maybe`', compiler/rustc_typeck/src/check/expr.rs:1309:13
stack backtrace:
   0:     0x7ff5b409e18d - std::backtrace_rs::backtrace::libunwind::trace::hb1109710b5bac759
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7ff5b409e18d - std::backtrace_rs::backtrace::trace_unsynchronized::h785364b6aed0029a
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7ff5b409e18d - std::sys_common::backtrace::_print_fmt::h74f64f39298dd76e
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7ff5b409e18d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h496e3b79162d5554
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7ff5b40f9f7c - core::fmt::write::hab33fa5859262fe7
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/core/src/fmt/mod.rs:1197:17
   5:     0x7ff5b408f801 - std::io::Write::write_fmt::h1710e3e0ee173be1
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/io/mod.rs:1672:15
   6:     0x7ff5b40a0e55 - std::sys_common::backtrace::_print::h2a4a41c4c8a478e3
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7ff5b40a0e55 - std::sys_common::backtrace::print::h04d400740890388f
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7ff5b40a0e55 - std::panicking::default_hook::{{closure}}::h33882b41ce9ace17
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/panicking.rs:295:22
   9:     0x7ff5b40a0b76 - std::panicking::default_hook::hed13b5f1fa769ff8
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/panicking.rs:314:9
  10:     0x7ff5b490f804 - rustc_driver[2985e0c6160588b5]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7ff5b40a152a - std::panicking::rust_panic_with_hook::hfb6f3749c685320d
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/panicking.rs:702:17
  12:     0x7ff5b40a1367 - std::panicking::begin_panic_handler::{{closure}}::h99f3f91cb6d654a6
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/panicking.rs:588:13
  13:     0x7ff5b409e644 - std::sys_common::backtrace::__rust_end_short_backtrace::h1d5304cdf1728a1c
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/sys_common/backtrace.rs:138:18
  14:     0x7ff5b40a1099 - rust_begin_unwind
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/panicking.rs:584:5
  15:     0x7ff5b4066263 - core::panicking::panic_fmt::hce2c0129c83bc4c4
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/core/src/panicking.rs:142:14
  16:     0x7ff5b40f6d28 - core::panicking::assert_failed_inner::h34a537bd3d0af8b9
  17:     0x7ff5b4cab2dc - core[1c8f1df5a59aa66f]::panicking::assert_failed::<rustc_typeck[9766a23a180a54e4]::check::diverges::Diverges, rustc_typeck[9766a23a180a54e4]::check::diverges::Diverges>
  18:     0x7ff5b5f0663d - <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_expr_kind
  19:     0x7ff5b4d3f886 - <core[1c8f1df5a59aa66f]::iter::adapters::map::Map<core[1c8f1df5a59aa66f]::ops::range::Range<usize>, <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::ArgMatrix>::new<<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_argument_types::{closure#3}>::{closure#0}::{closure#0}> as core[1c8f1df5a59aa66f]::iter::traits::iterator::Iterator>::fold::<(), core[1c8f1df5a59aa66f]::iter::traits::iterator::Iterator::for_each::call<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::Compatibility, <alloc[318c0fc51e3c1903]::vec::Vec<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::Compatibility> as alloc[318c0fc51e3c1903]::vec::spec_extend::SpecExtend<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::Compatibility, core[1c8f1df5a59aa66f]::iter::adapters::map::Map<core[1c8f1df5a59aa66f]::ops::range::Range<usize>, <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::ArgMatrix>::new<<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_argument_types::{closure#3}>::{closure#0}::{closure#0}>>>::spec_extend::{closure#0}>::{closure#0}>
  20:     0x7ff5b4d2c567 - <alloc[318c0fc51e3c1903]::vec::Vec<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::Compatibility> as alloc[318c0fc51e3c1903]::vec::spec_from_iter::SpecFromIter<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::Compatibility, core[1c8f1df5a59aa66f]::iter::adapters::map::Map<core[1c8f1df5a59aa66f]::ops::range::Range<usize>, <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::ArgMatrix>::new<<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_argument_types::{closure#3}>::{closure#0}::{closure#0}>>>::from_iter
  21:     0x7ff5b4d3f684 - <core[1c8f1df5a59aa66f]::iter::adapters::map::Map<core[1c8f1df5a59aa66f]::ops::range::Range<usize>, <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::ArgMatrix>::new<<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_argument_types::{closure#3}>::{closure#0}> as core[1c8f1df5a59aa66f]::iter::traits::iterator::Iterator>::fold::<(), core[1c8f1df5a59aa66f]::iter::traits::iterator::Iterator::for_each::call<alloc[318c0fc51e3c1903]::vec::Vec<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::Compatibility>, <alloc[318c0fc51e3c1903]::vec::Vec<alloc[318c0fc51e3c1903]::vec::Vec<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::Compatibility>> as alloc[318c0fc51e3c1903]::vec::spec_extend::SpecExtend<alloc[318c0fc51e3c1903]::vec::Vec<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::Compatibility>, core[1c8f1df5a59aa66f]::iter::adapters::map::Map<core[1c8f1df5a59aa66f]::ops::range::Range<usize>, <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::ArgMatrix>::new<<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_argument_types::{closure#3}>::{closure#0}>>>::spec_extend::{closure#0}>::{closure#0}>
  22:     0x7ff5b4d24f95 - <alloc[318c0fc51e3c1903]::vec::Vec<alloc[318c0fc51e3c1903]::vec::Vec<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::Compatibility>> as alloc[318c0fc51e3c1903]::vec::spec_from_iter::SpecFromIter<alloc[318c0fc51e3c1903]::vec::Vec<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::Compatibility>, core[1c8f1df5a59aa66f]::iter::adapters::map::Map<core[1c8f1df5a59aa66f]::ops::range::Range<usize>, <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::ArgMatrix>::new<<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_argument_types::{closure#3}>::{closure#0}>>>::from_iter
  23:     0x7ff5b4cb73f1 - <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::arg_matrix::ArgMatrix>::new::<<rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_argument_types::{closure#3}>
  24:     0x7ff5b5ed1726 - <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_argument_types
  25:     0x7ff5b5eaf244 - <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_call
  26:     0x7ff5b5f00b55 - <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_expr_kind
  27:     0x7ff5b5effbf0 - <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  28:     0x7ff5b5eed616 - <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_stmt
  29:     0x7ff5b5eeec57 - <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_block_with_expected
  30:     0x7ff5b5f00ec3 - <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_expr_kind
  31:     0x7ff5b5effbf0 - <rustc_typeck[9766a23a180a54e4]::check::fn_ctxt::FnCtxt>::check_expr_with_expectation
  32:     0x7ff5b5f31c06 - rustc_typeck[9766a23a180a54e4]::check::check::check_fn
  33:     0x7ff5b5fc63fe - <rustc_infer[fac6976779193def]::infer::InferCtxtBuilder>::enter::<&rustc_middle[e91eda85edc2d458]::ty::context::TypeckResults, <rustc_typeck[9766a23a180a54e4]::check::inherited::InheritedBuilder>::enter<rustc_typeck[9766a23a180a54e4]::check::typeck_with_fallback<rustc_typeck[9766a23a180a54e4]::check::typeck::{closure#0}>::{closure#1}, &rustc_middle[e91eda85edc2d458]::ty::context::TypeckResults>::{closure#0}>
  34:     0x7ff5b5f507ba - rustc_typeck[9766a23a180a54e4]::check::typeck
  35:     0x7ff5b6507984 - rustc_query_system[a80f8bd821f477c6]::query::plumbing::try_execute_query::<rustc_query_impl[bc93cb9e73dff18e]::plumbing::QueryCtxt, rustc_query_system[a80f8bd821f477c6]::query::caches::DefaultCache<rustc_span[8b48bd2bfbcc9443]::def_id::LocalDefId, &rustc_middle[e91eda85edc2d458]::ty::context::TypeckResults>>
  36:     0x7ff5b64ddcde - <rustc_query_impl[bc93cb9e73dff18e]::Queries as rustc_middle[e91eda85edc2d458]::ty::query::QueryEngine>::typeck
  37:     0x7ff5b5f5094f - rustc_typeck[9766a23a180a54e4]::check::typeck
  38:     0x7ff5b6507984 - rustc_query_system[a80f8bd821f477c6]::query::plumbing::try_execute_query::<rustc_query_impl[bc93cb9e73dff18e]::plumbing::QueryCtxt, rustc_query_system[a80f8bd821f477c6]::query::caches::DefaultCache<rustc_span[8b48bd2bfbcc9443]::def_id::LocalDefId, &rustc_middle[e91eda85edc2d458]::ty::context::TypeckResults>>
  39:     0x7ff5b64ddcde - <rustc_query_impl[bc93cb9e73dff18e]::Queries as rustc_middle[e91eda85edc2d458]::ty::query::QueryEngine>::typeck
  40:     0x7ff5b5fed5b8 - <rustc_middle[e91eda85edc2d458]::hir::map::Map>::par_body_owners::<rustc_typeck[9766a23a180a54e4]::check::typeck_item_bodies::{closure#0}>
  41:     0x7ff5b6da51fc - rustc_typeck[9766a23a180a54e4]::check::typeck_item_bodies
  42:     0x7ff5b709f75c - rustc_query_system[a80f8bd821f477c6]::query::plumbing::try_execute_query::<rustc_query_impl[bc93cb9e73dff18e]::plumbing::QueryCtxt, rustc_query_system[a80f8bd821f477c6]::query::caches::DefaultCache<(), ()>>
  43:     0x7ff5b70ccac1 - rustc_query_system[a80f8bd821f477c6]::query::plumbing::get_query::<rustc_query_impl[bc93cb9e73dff18e]::queries::typeck_item_bodies, rustc_query_impl[bc93cb9e73dff18e]::plumbing::QueryCtxt>
  44:     0x7ff5b6dcadf3 - <rustc_session[a46327cfa276e5ae]::session::Session>::time::<(), rustc_typeck[9766a23a180a54e4]::check_crate::{closure#7}>
  45:     0x7ff5b6dc1573 - rustc_typeck[9766a23a180a54e4]::check_crate
  46:     0x7ff5b6b33067 - rustc_interface[3f74620ae1096927]::passes::analysis
  47:     0x7ff5b7094c8f - rustc_query_system[a80f8bd821f477c6]::query::plumbing::try_execute_query::<rustc_query_impl[bc93cb9e73dff18e]::plumbing::QueryCtxt, rustc_query_system[a80f8bd821f477c6]::query::caches::DefaultCache<(), core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>>>
  48:     0x7ff5b70df01e - rustc_query_system[a80f8bd821f477c6]::query::plumbing::get_query::<rustc_query_impl[bc93cb9e73dff18e]::queries::analysis, rustc_query_impl[bc93cb9e73dff18e]::plumbing::QueryCtxt>
  49:     0x7ff5b6b16e37 - <rustc_interface[3f74620ae1096927]::passes::QueryContext>::enter::<rustc_driver[2985e0c6160588b5]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>>
  50:     0x7ff5b6b0111f - <rustc_interface[3f74620ae1096927]::interface::Compiler>::enter::<rustc_driver[2985e0c6160588b5]::run_compiler::{closure#1}::{closure#2}, core[1c8f1df5a59aa66f]::result::Result<core[1c8f1df5a59aa66f]::option::Option<rustc_interface[3f74620ae1096927]::queries::Linker>, rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>>
  51:     0x7ff5b6afa64f - rustc_span[8b48bd2bfbcc9443]::with_source_map::<core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>, rustc_interface[3f74620ae1096927]::interface::create_compiler_and_run<core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>, rustc_driver[2985e0c6160588b5]::run_compiler::{closure#1}>::{closure#1}>
  52:     0x7ff5b6b01fb2 - <scoped_tls[3b74a2cd32801712]::ScopedKey<rustc_span[8b48bd2bfbcc9443]::SessionGlobals>>::set::<rustc_interface[3f74620ae1096927]::interface::run_compiler<core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>, rustc_driver[2985e0c6160588b5]::run_compiler::{closure#1}>::{closure#0}, core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>>
  53:     0x7ff5b6b174bf - std[183c58237bb308d2]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[3f74620ae1096927]::util::run_in_thread_pool_with_globals<rustc_interface[3f74620ae1096927]::interface::run_compiler<core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>, rustc_driver[2985e0c6160588b5]::run_compiler::{closure#1}>::{closure#0}, core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>>::{closure#0}, core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>>
  54:     0x7ff5b6b176a9 - <<std[183c58237bb308d2]::thread::Builder>::spawn_unchecked_<rustc_interface[3f74620ae1096927]::util::run_in_thread_pool_with_globals<rustc_interface[3f74620ae1096927]::interface::run_compiler<core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>, rustc_driver[2985e0c6160588b5]::run_compiler::{closure#1}>::{closure#0}, core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>>::{closure#0}, core[1c8f1df5a59aa66f]::result::Result<(), rustc_errors[cc19d71e926465fe]::ErrorGuaranteed>>::{closure#1} as core[1c8f1df5a59aa66f]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  55:     0x7ff5b40ab463 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::ha4ec43e0ec331bcb
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/alloc/src/boxed.rs:1951:9
  56:     0x7ff5b40ab463 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h5e13b6a1911b918e
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/alloc/src/boxed.rs:1951:9
  57:     0x7ff5b40ab463 - std::sys::unix::thread::Thread::new::thread_start::h21c45ad5909aa02c
                               at /rustc/20a6f3a8a8ce5ae18d06b12cd7904bc5294ca753/library/std/src/sys/unix/thread.rs:108:17
  58:     0x7ff5b3c8c54d - <unknown>
  59:     0x7ff5b3d11874 - clone
  60:                0x0 - <unknown>

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.64.0-nightly (20a6f3a8a 2022-06-25) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `main`
#1 [typeck] type-checking `main::{closure#0}`
#2 [typeck_item_bodies] type-checking all item bodies
#3 [analysis] running analysis passes on this crate
end of query stack
```

</p>
</details>

### PR DESCRIPTION
**Pull request by compiler-errors**
Fixes a weird suggestion in #98784

found later:
Fixes #98894
Fixes #98897

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? @jackh726 do you mind taking a look at this? it actually ends up making the code slightly simpler, too.

this ICEs in production found by @WaffleLapkin, so I'm also gonna beta nominate this since it's a pretty simple to understand and self-contained fix.

**Comment by compiler-errors** [comment]
well, I guess it didn't ICE in the motivating example that Waffle originally found... but idk. Worth nominating at least. Hopefully this is the last of the arg mismatch algorithm issues, hehe.

File: compiler/rustc_typeck/src/check/fn_ctxt/checks.rs [hunk 1]
```diff
@@ -499,15 +512,9 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
             }
 
-            let provided_arg: &hir::Expr<'tcx> = &provided_args[provided_idx];
-            let expectation = Expectation::rvalue_hint(self, expected_input_ty);
-            // FIXME: check that this is safe; I don't believe this commits any of the obligations, but I can't be sure.
-            //
-            //   I had another method of "soft" type checking before,
-            //   but it was failing to find the type of some expressions (like "")
-            //   so I prodded this method and made it pub(super) so I could call it, and it seems to work well.
-            let checked_ty = self.check_expr_kind(provided_arg, expectation);
```
[the remark this answers is not in the document]
Comment (compiler-errors) [reply compiler/rustc_typeck/src/check/fn_ctxt/checks.rs:509]: Good catch.

File: src/test/ui/issues/issue-3044.rs [hunk 2]
```diff
@@ -1,7 +1,6 @@
 fn main() {
     let needlesArr: Vec<char> = vec!['a', 'f'];
     needlesArr.iter().fold(|x, y| {
-    });
-    //~^^ ERROR mismatched types
-    //~| ERROR this function takes 2 arguments but 1 argument was supplied
+    //~^ ERROR this function takes 2 arguments but 1 argument was supplied
+});
```
Comment (estebank) [inline src/test/ui/issues/issue-3044.rs:5]: Incorrect formatting
```suggestion
    });
```

**Comment by estebank** [comment]
r=me after rebase

**Comment by Mark-Simulacrum** [comment]
@compiler-errors it looks like this doesn't backport to the beta branch cleanly. Can you prepare a beta-branch targeted PR with this PR's changes that we can review and merge?

**Comment by compiler-errors** [comment]
I can do that. Thanks for the heads up.

**Comment by compiler-errors** [comment]
Oh, yeah, this definitely doesn't cleanly backport. 

For some reason I had thought that the PR this is building off (#97542) was landing in 1.63, not 1.64. I will have to use a different solution for 1.63.

**Pull request merged**